### PR TITLE
[CMSIS-NN] Fix scalar to tensor constant pass when checked type is TupleType

### PR DESCRIPTION
**Pull request by lhutton1**
A failure is seen when compiling some networks due to assumption checked_type is a TensorTypeNode: 

```Check failed: (node != nullptr) is false: Expected type to be relay.TensorType, but get TupleType```

Fixed by reshuffling some of the skip conditions and also ensures arguments are of type VarNode or ConstantNode (where necessary) before casting.

cc @ashutosh-arm @Mousius @manupa-arm

### CODE REVIEW

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc [hunk 1]
```diff
@@ -117,21 +117,19 @@ class ScalarToTensorConstantMutator : public MixedModeMutator {
...
       Expr arg = call->args[i];
       new_args.push_back(arg);
-      if (!arg->checked_type_.defined()) {
-        continue;
-      }
-      auto* arg_type = arg->type_as<TensorTypeNode>();
-      if (arg_type->shape.size() != 0 || arg.as<ConstantNode>()) {
+      const auto* arg_var = arg.as<VarNode>();
+      const auto* arg_type = arg->checked_type_.as<TensorTypeNode>();
+      if (!arg_var || !arg_type || arg_type->shape.size() != 0) {
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:122]: nit : could use "->IsInstance<VarNode>()" if arg_var and arg_type is not used anymore.

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc [hunk 2]
```diff
@@ -117,21 +117,19 @@ class ScalarToTensorConstantMutator : public MixedModeMutator {
...
       int tensor_arg_id = (i + 1) % 2;
       Expr tensor_arg = call->args[tensor_arg_id];
-      if (!tensor_arg->checked_type_.defined()) {
+      const auto* tensor_type = tensor_arg->checked_type_.as<TensorTypeNode>();
+      if (!tensor_type) {
         continue;
       }
-      TensorType tensor_type = GetRef<TensorType>(tensor_arg->type_as<TensorTypeNode>());
-      new_args.Set(i, Var(arg_name, tensor_type));
+      new_args.Set(i, Var(arg_name, GetRef<TensorType>(tensor_type)));
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:132]: nit : same here we dont need to extract the node just to covert it back to Ref

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc [hunk 3]
```diff
@@ -142,19 +140,18 @@ class ScalarToTensorConstantMutator : public MixedModeMutator {
     for (uint32_t i = 0; i < call->args.size(); ++i) {
       new_args.push_back(call->args[i]);
       Expr scalar_arg = call->args[i];
-      if (!scalar_arg->checked_type_.defined()) {
-        continue;
-      }
-      Array<PrimExpr> scalar_shape = scalar_arg->type_as<TensorTypeNode>()->shape;
-      if (scalar_shape.size() != 0 || scalar_arg.as<ConstantNode>() == nullptr) {
+      const auto* scalar_const = scalar_arg.as<ConstantNode>();
+      const auto* scalar_type = scalar_arg->checked_type_.as<TensorTypeNode>();
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:144]: same nit as above

File: src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc [hunk 4]
```diff
@@ -142,19 +140,18 @@ class ScalarToTensorConstantMutator : public MixedModeMutator {
...
       }
       int tensor_arg_id = (i + 1) % 2;
       Expr tensor_arg = call->args[tensor_arg_id];
-      if (!tensor_arg->checked_type_.defined()) {
+      const auto* tensor_type_node = tensor_arg->checked_type_.as<TensorTypeNode>();
+      if (!tensor_type_node) {
         continue;
       }
-      TensorType tensor_type = GetRef<TensorType>(tensor_arg->type_as<TensorTypeNode>());
+      TensorType tensor_type = GetRef<TensorType>(tensor_type_node);
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/scalar_to_tensor_constant.cc:154]: same nit as above

**Review by manupa-arm — COMMENTED** [review]
Thanks @lhutton1 !

Left some few nits to make the code a bit leaner (a possible candidate for a follow up! as this is a bug fix).

I can see from the test that fix is working, however I am bit struggling to understand how tuples are handled in the code (surely it seems to be handled)

**Comment by lhutton1** [comment]
Thanks @manupa-arm, great points! Taking a second look at this I believe I need to investigate further, the behavior of my changes will mean tuples are simply skipped over, which I don't believe is correct.

**Comment by lhutton1** [comment]
After further investigation, the bug was found to be caused by the pass running on all functions in the input graph, not just CMSIS-NN functions. Therefore closing since the bug fix in #10375 supersedes this.

**Pull request closed without merging**